File: frontend/src/types/recommendation.ts
import type { ApiCategoryRef, ApiCollectionRef, ApiItemListItem, ApiItemStatus } from "./api";

export type RecommendationStatusFilter = "PLANNED" | "COMPLETED" | "ALL";
export type RecommendationCandidateType = "ITEM" | "COLLECTION";

export interface RandomRecommendationRequest {
  category_id: string;
  status_filter: RecommendationStatusFilter;
}

export interface RandomRecommendationResponse {
  category: ApiCategoryRef;
  status_filter: RecommendationStatusFilter;
  candidate_type: RecommendationCandidateType | null;
  candidate_id: string | null;
  collection: ApiCollectionRef | null;
  items: ApiItemListItem[];
  eligible_candidate_count: number;
}

export interface CreateRecommendationHistoryRequest {
  category_id: string;
  status_filter: RecommendationStatusFilter;
  candidate_type: RecommendationCandidateType;
  candidate_id: string;
}

export interface RecommendationHistoryListItem {
  id: string;
  category: ApiCategoryRef;
  status_filter: RecommendationStatusFilter;
  collection: ApiCollectionRef | null;
  selected_at: string;
  title: string;
  item_count: number;
  poster_url: string | null;
  release_year: number | null;
  candidate_type: RecommendationCandidateType;
}

export interface RecommendationHistoryPage {
  items: RecommendationHistoryListItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
  has_next: boolean;
  has_previous: boolean;
}

export interface RecommendationHistoryDetailItem {
  item_id: string;
  title_snapshot: string;
  status_at_selection: ApiItemStatus;
  sort_order: number;
  item_exists: boolean;
  current: ApiItemListItem | null;
}

export interface RecommendationHistoryDetail {
  id: string;
  category: ApiCategoryRef;
  status_filter: RecommendationStatusFilter;
  collection: ApiCollectionRef | null;
  selected_at: string;
  candidate_type: RecommendationCandidateType;
  items: RecommendationHistoryDetailItem[];
}

export interface RecommendationHistoryDeleteAllResponse {
  deleted_count: number;
}

File: backend/app/schemas/recommendation.py
"""Schemas for random recommendation and recommendation history (REC-1)."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.models import ItemStatus, StatusFilter
from app.schemas import CategoryRef, CollectionRef, ItemListItem

RecommendationCandidateType = Literal["ITEM", "COLLECTION"]


class RandomRecommendationRequest(BaseModel):
    category_id: UUID
    status_filter: StatusFilter


class RandomRecommendationResponse(BaseModel):
    category: CategoryRef
    status_filter: StatusFilter
    candidate_type: RecommendationCandidateType | None
    candidate_id: UUID | None = None
    collection: CollectionRef | None = None
    items: list[ItemListItem]
    eligible_candidate_count: int


class CreateRecommendationHistoryRequest(BaseModel):
    category_id: UUID
    status_filter: StatusFilter
    candidate_type: RecommendationCandidateType
    candidate_id: UUID


class RecommendationHistoryListItem(BaseModel):
    id: UUID
    category: CategoryRef
    status_filter: StatusFilter
    collection: CollectionRef | None
    selected_at: datetime
    title: str
    item_count: int
    poster_url: str | None = None
    release_year: int | None = None
    candidate_type: RecommendationCandidateType


class RecommendationHistoryListResponse(BaseModel):
    items: list[RecommendationHistoryListItem]
    page: int
    page_size: int
    total: int
    total_pages: int
    has_next: bool
    has_previous: bool


class RecommendationHistoryDetailItem(BaseModel):
    item_id: UUID
    title_snapshot: str
    status_at_selection: ItemStatus
    sort_order: int
    item_exists: bool = True
    current: ItemListItem | None = None


class RecommendationHistoryDetailResponse(BaseModel):
    id: UUID
    category: CategoryRef
    status_filter: StatusFilter
    collection: CollectionRef | None
    selected_at: datetime
    candidate_type: RecommendationCandidateType
    items: list[RecommendationHistoryDetailItem]


class RecommendationHistoryDeleteAllResponse(BaseModel):
    deleted_count: int = Field(ge=0)
